components/server/src/auth/resource-access.ts

```diff
@@ -0,0 +1,266 @@
+import { Workspace, WorkspaceInstance, User, Snapshot, GitpodToken, Token } from "@gitpod/gitpod-protocol";
+import { injectable } from "inversify";
+
+declare var resourceInstance: GuardedResource;
+export type GuardedResourceKind = typeof resourceInstance.kind;
+
+export type GuardedResource =
+    GuardedWorkspace |
+    GuardedWorkspaceInstance |
+    GuardedUser |
+    GuardedSnapshot |
+    GuardedGitpodToken |
+    GuardedToken |
+    GuardedUserStorage
+;
+
+export interface GuardedWorkspace {
+    kind: "workspace";
+    subject: Workspace;
+}
+
+export interface GuardedWorkspaceInstance {
+    kind: "workspaceInstance";
+    subject: WorkspaceInstance | undefined;
+    workspaceOwnerID: string;
+}
+
+export interface GuardedUser {
+    kind: "user";
+    subject: User;
+}
+
+export interface GuardedSnapshot {
+    kind: "snapshot";
+    subject: Snapshot | undefined;
+    workspaceOwnerID: string;
+}
+
+export interface GuardedUserStorage {
+    kind: "userStorage";
+    userID: string;
+    uri: string;
+}
+
+export interface GuardedGitpodToken {
+    kind: "gitpodToken";
+    subject: GitpodToken;
+}
+
+export interface GuardedToken {
+    kind: "token";
+    subject: Token;
+    tokenOwnerID: string;
+}
+
+export type ResourceAccessOp =
+    "create" |
+    "update" |
+    "get"    |
+    "delete"
+;
+
+export const ResourceAccessGuard = Symbol("ResourceAccessGuard");
+
+export interface ResourceAccessGuard {
+    canAccess(resource: GuardedResource, operation: ResourceAccessOp): Promise<boolean>;
+}
+
+
+export interface WithResourceAccessGuard {
+    resourceGuard?: ResourceAccessGuard;
+}
+
+/**
+ * CompositeResourceAccessGuard grants access to resources if at least one of its children does.
+ */
+export class CompositeResourceAccessGuard implements ResourceAccessGuard {
+    
+    constructor(protected readonly children: ResourceAccessGuard[]) {}
+
+    async canAccess(resource: GuardedResource, operation: ResourceAccessOp): Promise<boolean> {
+        // if a single guard permitts access, we're good to go
+        return (await Promise.all(this.children.map(c => c.canAccess(resource, operation)))).some(x => x);
+    }
+
+}
+
+/**
+ * OwnerResourceGuard grants access to resources if the user asking for access is the owner of that
+ * resource.
+ */
+@injectable()
```

Comment:
not sure that it should be injectable? it seems that instances are created with `new` operator